protocols/kad/src/behaviour.rs

```diff
@@ -524,7 +524,11 @@ where
     ///
     /// If the routing table has been updated as a result of this operation,
     /// a [`KademliaEvent::RoutingUpdated`] event is emitted.
-    pub fn add_address(&mut self, peer: &PeerId, address: Multiaddr) -> RoutingUpdate {
+    pub fn add_address(
+        &mut self,
+        peer: &PeerId,
+        address: Multiaddr,
+    ) -> Result<RoutingUpdateOk, RoutingUpdateError> {
```

Comment:
Sorry that this is only coming up now but I just realized that we have a feature coming up that conflicts with this idea. It is here: https://github.com/libp2p/rust-libp2p/pull/4371

This function will likely be deprecated in favor of a general `add_address_of_peer` function on `Swarm`. This means, we won't be able to return a specific error there!

In fact, we won't be able to return anything at all.

Thus, I think we should actually change this PR into preparing this API for the above change. Here is what I'd suggest but I am interested to hear your and @mxinden's opinion:

1. Mark `RoutingUpdate` as deprecated but retain the current functionality.
2. Always emit `KademliaEvent::RoutingUpdated`, even for the immediate `RoutingUpdate::Success` case
3. Add a `debug_assert` to the `SelfEntry` case. If we attempt to add ourselves to the routing table, this is a programmer bug and should not happen. We can ignore that in production (but should probably emit a fat `warn!`), hence the `debug_assert`.
4. We already log a warning for the bucket-full case so we can just leave it at that.

This makes this a backwards-compatible change and we can land it right away.

Reply:
I missed this as well. I am sorry @arsenron.

Suggestion overall sounds good to me.

> Add a `debug_assert` to the `SelfEntry` case. If we attempt to add ourselves to the routing table, this is a programmer bug and should not happen. We can ignore that in production (but should probably emit a fat `warn!`), hence the `debug_assert`.

Panicing on invalid user input seems rather drastic. Wouldn't a `warn!` suffice in all cases? Also, not as part of this pull request, we should handle the `SelfEntry` on the `libp2p-swarm` level. We should at least log when a `NetworkBehaviour` or user (via `Swarm::add_address_of_peer`) wants to report a an external address of a remote peer where that peer is actually the local peer.

Reply:
I missed this as well. I am sorry @arsenron.

Suggestion overall sounds good to me.

> Add a `debug_assert` to the `SelfEntry` case. If we attempt to add ourselves to the routing table, this is a programmer bug and should not happen. We can ignore that in production (but should probably emit a fat `warn!`), hence the `debug_assert`.

Panicing on invalid user input seems rather drastic. Wouldn't a `warn!` suffice in all cases? Also, not as part of this pull request, we should handle the `SelfEntry` on the `libp2p-swarm` level. We should at least log when a `NetworkBehaviour` or user (via `Swarm::add_address_of_peer`) wants to report a an external address of a remote peer where that peer is actually the local peer.

Reply:
> I missed this as well. I am sorry @arsenron.
> 
> Suggestion overall sounds good to me.
> 
> > Add a `debug_assert` to the `SelfEntry` case. If we attempt to add ourselves to the routing table, this is a programmer bug and should not happen. We can ignore that in production (but should probably emit a fat `warn!`), hence the `debug_assert`.
> 
> Panicing on invalid user input seems rather drastic. Wouldn't a `warn!` suffice in all cases?

The warn will be there in all cases but pancking in debug builds is fine, no? You didn't validate your inputs correctly in that case.

> Also, not as part of this pull request, we should handle the `SelfEntry` on the `libp2p-swarm` level. We should at least log when a `NetworkBehaviour` or user (via `Swarm::add_address_of_peer`) wants to report a an external address of a remote peer where that peer is actually the local peer.

True, as part of #4371, we should probably not forward any of these events! cc @StemCll

Reply:
@arsenron I fixed the merge conflicts for you. Are you still interested in working on this given the above insight?

Reply:
@thomaseizinger sorry, little busy now, but I will check it a bit later